### PR TITLE
SNOW-508730  Introduce new config: processing.guarantee

### PR DESCRIPTION
**Pull request by sfc-gh-anwang**
https://snowflakecomputing.atlassian.net/browse/SNOW-508731
`processing.guarantee=at_least_once/exactly_once`
Allow enable exactly once semantics property in the Kafka connector configuration file

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java [hunk 1]
```diff
@@ -491,4 +526,41 @@ public String toString() {
       return name().toLowerCase(Locale.ROOT);
     }
   }
+
+  /**
+   * Enum which represents the type of processing guarantees that the customer want (either
+   * at_least_once (default) or exactly_once
+   */
+  public enum ProcessingGuarantee {
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:534]: Should this be `IngestionProcessingGuarantee`

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java [hunk 2]
```diff
@@ -491,4 +526,41 @@ public String toString() {
...
+    AT_LEAST_ONCE,
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:535]: Can we add java docs for each of the enum types.

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java [hunk 3]
```diff
@@ -139,6 +141,11 @@ public void start(final Map<String, String> parsedConfig) {
           Boolean.parseBoolean(parsedConfig.get(SnowflakeSinkConnectorConfig.JMX_OPT));
     }
 
+    // Get the processing guarantee type from config, default to at_least_once
+    SnowflakeSinkConnectorConfig.ProcessingGuarantee processingGuarantee =
+        SnowflakeSinkConnectorConfig.ProcessingGuarantee.of(
+            parsedConfig.getOrDefault(PROCESSING_GUARANTEE, null));
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java:147]: Lets keep the default to `AT_LEAST_ONCE`

File: src/test/java/com/snowflake/kafka/connector/ConnectorConfigTest.java [hunk 4]
```diff
@@ -268,4 +268,27 @@ public void testJMX_invalid_value() {
...
+    Utils.validateConfig(config);
+
+    config.put(SnowflakeSinkConnectorConfig.PROCESSING_GUARANTEE, "exactly_once");
+    Utils.validateConfig(config);
+
+    config.put(SnowflakeSinkConnectorConfig.PROCESSING_GUARANTEE, "");
+    Utils.validateConfig(config);
+
+    config.put(SnowflakeSinkConnectorConfig.PROCESSING_GUARANTEE, null);
+    Utils.validateConfig(config);
```
Comment (sfc-gh-japatel) [inline src/test/java/com/snowflake/kafka/connector/ConnectorConfigTest.java:285]: are empty and null fine?

File: src/main/java/com/snowflake/kafka/connector/Utils.java [hunk 5]
```diff
@@ -499,6 +500,18 @@ static String validateConfig(Map<String, String> config) {
       }
     }
 
+    try {
+      SnowflakeSinkConnectorConfig.ProcessingGuarantee.of(
+          config.getOrDefault(PROCESSING_GUARANTEE, null));
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/Utils.java:505]: What will be `enum.of(null)`? Will it be allowed?

**Review by sfc-gh-japatel — CHANGES_REQUESTED** [review]
left some comments. Overall looks good, will wait for one more round.

File: src/main/java/com/snowflake/kafka/connector/Utils.java (hunk 5 above)
Comment (sfc-gh-japatel) [reply src/main/java/com/snowflake/kafka/connector/Utils.java:505]: Ignore, I checked the method. Looks good, but lets keep default to `AT_LEAST_ONCE`

**Review by sfc-gh-japatel — APPROVED** [review]
minor changes, lgtm!

File: src/main/java/com/snowflake/kafka/connector/Utils.java (hunk 5 above)
Comment (sfc-gh-anwang) [reply src/main/java/com/snowflake/kafka/connector/Utils.java:505]: Yah, the enum.of(null) currently will return default AT_LEAST_ONCE. It also makes sense to throw exception when user provide an empty string, should i update the logic?

File: src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java (hunk 1 above)
Comment (sfc-gh-anwang) [reply src/main/java/com/snowflake/kafka/connector/SnowflakeSinkConnectorConfig.java:534]: Yah, I can update the naming

**Review by sfc-gh-japatel — APPROVED** [review]
lgtm

**Pull request merged**